driver/src/main/scala/core/nodeset/ChannelFactory.scala

```diff
@@ -91,29 +102,18 @@ private[reactivemongo] final class ChannelFactory(
 
       debug(s"Created new channel #${channel.id} to ${host}:${port} (registered = ${channel.isRegistered})")
 
-      if (channel.isRegistered) {
-        initChannel(channel, host, port, receiver)
-      } else {
-        // Set state as attributes so available for the coming init
-        channel.attr(ChannelFactory.hostKey).set(host)
-        channel.attr(ChannelFactory.portKey).set(port)
-        channel.attr(ChannelFactory.actorRefKey).set(receiver)
-      }
+      // Set state as attributes so available for the coming init
+      attrsHolder ! AddChannelSettings(channel.id(), host, port, receiver)
```

Comment:
- I don't see how this would be a fix?
- Anyway why using an extra actor to pass the same values as current code (which was working with channel attributes / simpler)?

Reply:
the problem was `initChannel()` was being called with only some of the attributes set (at the same time they were being set, actually).

this code sends the attributes to the actor, and if they haven't made it yet when the requests to fetch the attributes get there, the fetch is stashed until the attributes come in, then the request is unstashed and replied to. so order no longer matters.

the old code would often have an entry in the logs that it was skipping init because host was null. well, actorRef and port had the same issue (and actorRef being null would cause an NPE in `MongoHandler`). On my machine, many of my Channels would fail, so even when the connection worked, there was often delays and errors getting there. Sometimes the first attempt wouldn't work, and after a retry they would, with more errors logged. After this change, all connections are working the first time, and quickly/error-free. 

An actor isn't the only way to accomplish this, of course, but something has to be done to make sure that the host/port/receiver are available at the time that `initChannel()` is called.

Reply:
Here's one test run on 0.18.4: https://pastebin.com/hnTEkcDb

19 tries, the last one failed and threw the NPE because the receiver was null. on 13 of them you can see the line `Skip channel init as host is null`, showing that by the time `initChannel()` was called, none of the attributes had been set yet. So the NPE (last) case is the most insidious, because `host` HAD been set, `port` may or may not have been set, and `receiver` had NOT.

Reply:
I mean think about it this way...

The way the code wants the world to work is:
create() is called
we call channelFactory.connect()
we set 3 attributes on the channel
netty calls initChannel(), we fetch the 3 attributes from the channel, do initialization, everything's great

The way it actually works on a fast machine with 12 threads of execution:
create() is called
we call channelFactory.connect()
netty starts initializing the connection in another thread
we start setting attributes on the channel on thread 1
netty's tries to initialize on thread 2
the attributes are thread-safe! great!
thread 1 write host
thread 1 write port
thread 2 read host
thread 2 read port
thread 2 read receiver
thread 1 write receiver
NullPointerException when receiver is used

the attributes are thread-safe, sure, but there are legitimate timing issues here. my first thought was synchronization on the channel factory, but i'd much rather use an actor and keep things actually running on multiple threads instead of serializing all channel creation

Reply:
There are two points:

- When it's required to pass the attributes.
- How to pass the attributes.

Anyway on the former point, using an actor is over-engineered.

The first point need to be check, as previous reproducer was not showing result locally.

Reply:
you can reproduce this easily.

it's checking for isRegistered, so it's not always going to be registered at that point, right? it's an asynchronous event. so sometimes it will be registered by the time you check it, and sometimes it won't. so for the sake of argument, drop that code out, and always use the other method (passing attributes). now, before setting the attributes, add a `Thread.sleep(5000)`. this guarantees that that code won't be run first (unlike now, where sometimes it is, and sometimes it isn't. the nature of multithreading). you will see the same errors i'm having. even better, add a Thread.sleep before setting the actorRefKey, make sure trace is on for MongoHandler, and you'll see the NPEs yourself (because init will have been called with a valid host and port, but no receiver).

if you want, i can put debug lines in a branch, showing the order of operations, and showing you from debug output on a fast machine that the order is random, and that it very much matters.

anyway, however you want me to prove to you that this is an issue, I will.

as for being over-engineered, i'm open to suggestions. the only other things i can think of is synchronizing on create() and initChannel(), making sure only one channel can be created or initted at a time, or removing the isRegistered check and set up a loop in initChannel with sleeps if host or actorRef haven't been set yet, which is terrible and hacky. i mean, akka is here to solve concurrency issues, why not use it. but again, open to any suggestions.

Reply:
> it's checking for isRegistered, so it's not always going to be registered at that point, right? it's an asynchronous event. so sometimes it will be registered by the time you check it, and sometimes it won't. so for the sake of argument, drop that code out, and always use the other method (passing attributes).

If it's properly registered by Netty at this time, the internal `initChannel` is called by just passing the create arguments (safe), where there is no need to set the attributes. Otherwise it comes to the case where it's needed to set the attributes.

> now, before setting the attributes, add a `Thread.sleep(5000)`. this guarantees that that code won't be run first (unlike now, where sometimes it is, and sometimes it isn't. the nature of multithreading). 

(I think you can avoid the phrasing such as I'm not aware what is multithreading)

> you will see the same errors i'm having. even better, add a Thread.sleep before setting the actorRefKey, make sure trace is on for MongoHandler, and you'll see the NPEs yourself (because init will have been called with a valid host and port, but no receiver).
> 
> if you want, i can put debug lines in a branch, showing the order of operations, and showing you from debug output on a fast machine that the order is random, and that it very much matters.
> 
> anyway, however you want me to prove to you that this is an issue, I will.
> 
> as for being over-engineered, i'm open to suggestions. the only other things i can think of is synchronizing on create() and initChannel(), making sure only one channel can be created or initted at a time, or removing the isRegistered check and set up a loop in initChannel with sleeps if host or actorRef haven't been set yet, which is terrible and hacky.

I'm not convinced that `Try(Await.result((attrsHolder ? GetChannelSettings(channel.id())).mapTo[ChannelSettings], Duration.Inf))` is the safer/least "hacky" way/

> i mean, akka is here to solve concurrency issues, why not use it.

Because it has a cost, both of maintenance (if unnecessary complexity) and at runtime.

Will have a look. Thanks for your time.

Reply:
Sorry, I truly didn't mean to sound insulting, i just tend to go off and start talking about something at length. wasn't trying to tell you anything.

You're right, any time i use `Await` i think there's a problem...but netty doesn't have an asynchronous channel initializer, as far as i know, and i saw `Await`s in the code yesterday, like in `MongoDBSystem.connectAll`, so I figured using it at the barrier between sync & async code was an ok place. 

Definitely agree simpler code is better, and I didn't LIKE the actor solution, just couldn't think of a better one. I'd disagree a bit with the runtime cost, it's really just a wrapper around a `ConcurrentLinkedQueue` (I'm sure you know this), and a single actor is super cheap (also sure you know that).

Anyway, thanks, I hope you can come up with a better/simpler way of doing this. if you just have an idea of a better way but don't have time to deal with it, i'm happy to redo the implementation

Reply:
> ... any time i use `Await` i think there's a problem...but netty doesn't have an asynchronous channel initializer, as far as i know, and i saw `Await`s in the code yesterday, like in `MongoDBSystem.connectAll`, 

There is no infinite await in the db system, but some with explicit heartbeat frequency/connection timeout, according chosen option, with handle timeout.

> so I figured using it at the barrier between sync & async code was an ok place.
> 
> Definitely agree simpler code is better, and I didn't LIKE the actor solution, just couldn't think of a better one. I'd disagree a bit with the runtime cost, it's really just a wrapper around a `ConcurrentLinkedQueue` (I'm sure you know this), and a single actor is super cheap (also sure you know that).

I think that's an actor is a little bit more that a queue, but even a concurrent queue won't be the first solution to come to my mind to work around this issue.

> 
> Anyway, thanks, I hope you can come up with a better/simpler way of doing this. if you just have an idea of a better way but don't have time to deal with it, i'm happy to redo the implementation

Something as been push using netty factory default attributes, that ensure attributes are set even before `connect` (before the beginning of the async lifecycle).

Reply:
yeah i just saw your code, i didn't even think about that, and didn't realize that channelfactories were cheap to create.

thanks!